topics/tutorial-first-steps-with-kotlin-rpc.topic

```diff
@@ -0,0 +1,600 @@
+            languages and are serialized using a compact and efficient binary format. Microservice frameworks like
+            Quarkus and Micronaut already have support for gRPC.
+        </p>
+        <p>
+            It would be difficult for Kotlin RPC to compete with gRPC, and there would be no benefit in this for the
+            Kotlin community. Thankfully, there are no plans to do this. Instead, the intention is for kotlinx.rpc to be
+            compatible and interoperable with gRPC. It will be possible for kotlinx.rpc services to use gRPC as their
+            networking protocol and for kotlinx.rpc clients to call gRPC services. kotlin.rpc will use <a
+                href="https://kotlin.github.io/kotlinx-rpc/transport.html">its own kRPC protocol</a> as the default
+            option (as is the case in our current example), but there will be nothing to prevent you from choosing gRPC
+            instead.
+        </p>
+    </chapter>
+    <chapter title="Extend the example" id="extend-the-example">
+        <p>
+            Finally, let's enhance the complexity of our example application to establish a solid foundation for future
+            development.
+        </p>
+        <procedure id="extend-server">
+            <step>
+                In the
+                <path>PizzaShop.kt</path>
+                file in each project, extend the <code>orderPizza</code> method by including the client’s ID, and add a
+                <code>viewOrders</code> method that returns all the pending orders for a specified client:
+                <code-block lang="kotlin"
+                            src="snippets/tutorial-kotlin-rpc-server/src/main/kotlin/model/PizzaShop.kt"
+                            include-lines="1-5,9-11"
+                />
+                <p>
+                    This new version of the interface will need to be shared in both the client and server projects. You
+                    can take advantage of the coroutines library by returning a <code>Flow</code> rather than a <code>List</code>
+                    or <code>Set</code>. This will allow us to steam the information to the client one pizza at a time.
+                </p>
+            </step>
+            <step>
+                On the server side, navigate to the
+                <path>Main.kt</path>
+                file and implement this functionality by storing the current orders in a map of lists:
+                <code-block lang="kotlin"
+                            src="snippets/tutorial-kotlin-rpc-server/src/main/kotlin/Main.kt"
+                            include-lines="18-20,26-49"
+                />
+                <p>
+                    Note that this implementation assumes the same <code>PizzaShopImpl</code> object will be shared
+                    across all clients, and ignores the thread safety issues this would cause. This will be addressed in
```

Comment:
Actually one `PizzaShopImpl` instance is created per one client instance. Still not thread-safe (at least for now), but the model is a bit different.

Reply:
So is this accurate:
One `PizzaShopImpl` instance is created for each client instance. This avoids conflicts between clients but does not address thread safety within a single client's instance.

Reply:
Yes. I just want to clarify the terminology a little. We have a client as `RPCClient`, which communicates with the `RPCServer`. Each client can have multiple services attached to it, even of the same type. This means that each of these services would have its own `Impl` instance on this server. So for two `PizzaShop` services where will be two different `PizzaShopImpl` instances on a server:
```kotlin
val service1 = client.withService<PizzaShop>()
val service2 = client.withService<PizzaShop>()

service1 != service2
```

So to sum up, saying `One PizzaShopImpl instance is created for each client instance` is not wrong, but `for each client service instance` would eliminate all ambiguity. 

This a technical part, which option is better to read and understand for a user is not obvious to me, so I would like to hear your opinion on it